// mcp_constants/index.js
var termStatus = require("./TermStatus.js");

var exports = {};
exports.termStatus = termStatus;
exports.userType = require("./UserType.js");
exports.notifyType = require("./NotifyType.js");
exports.gameType = require("./GameType.js");
exports.msgType = require("./MsgType.js");
exports.msgStatus = require("./MsgStatus.js");
exports.ticketStatus = require("./TicketStatus.js");
exports.ticketPrintStatus = require("./TicketPrintStatus.js");
exports.ticketPrintQueenStatus = require("./TicketPrintQueenStatus.js");
exports.notifyStatus = require("./NotifyStatus.js");
exports.gameGrade = require("./GameGrade.js");
exports.termReportType = require('./TermReportType.js');
exports.customerStatus = require('./CustomerStatus.js');
exports.digestType = require('./DigestType.js');
exports.errUrl = require('./ErrUrl.js');
exports.accountSelect = require('./AccountSelect.js');
exports.gameStatus = require('./GameStatus.js');
exports.termStatusForPage = require('./TermStatusForPage.js');


module.exports = exports;
