// mcp_control/PageControl.js
var dc = require('mcp_db').dc;
var esut = require('easy_util');
var digestUtil = esut.digestUtil;
var adminPageControl = require("./AdminPageControl.js");
var channelPageControl = require("./ChannelPageControl.js");
var termPageControl = require("./TermPageControl.js");
var gamePageControl = require("./GamePageControl.js");
var msgPageControl = require("./MsgPageControl.js");
var customerPageControl = require("./CustomerPageControl.js");
var sysPageControl = require("./SysPageControl.js");
var gameGradePageControl = require("./GameGradePageControl.js");
var moneyLogPageControl = require("./MoneyLogPageControl.js");
var orderPageControl = require("./OrderPageControl.js");
var ticketPageControl = require("./TicketPageControl.js");
var relationPageControl = require("./RelationPageControl.js");
var printPageControl = require("./PrintPageControl.js");
var notifyPageControl = require("./NotifyPageControl.js");
var mongoPageControl = require("./MongoPageControl.js");
var mysqlPageControl = require("./MysqlPageControl.js");
var reportPageControl = require('./ReportPageControl.js');
var rechargePageControl = require('./RechargePageControl.js');
var salePageControl = require('./SalePageControl.js');

var PageControl = function(){};

PageControl.prototype.handle = function(headNode, bodyNode, cb)
{
    console.log(bodyNode);
    var self = this;
    var cmd = headNode.cmd;
    self[cmd[0]](headNode, bodyNode, cb);
};

PageControl.prototype.admin = function(headNode, bodyNode, cb)
{
    adminPageControl.handle(headNode, bodyNode, cb);
};


PageControl.prototype.game = function(headNode, bodyNode, cb)
{
    gamePageControl.handle(headNode, bodyNode, cb);
};


PageControl.prototype.term = function(headNode, bodyNode, cb)
{
    termPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.msg = function(headNode, bodyNode, cb)
{
    msgPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.customer = function(headNode, bodyNode, cb)
{
    customerPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.channel = function(headNode, bodyNode, cb)
{
    channelPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.sys = function(headNode, bodyNode, cb)
{
    sysPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.gamegrade = function(headNode, bodyNode, cb)
{
    gameGradePageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.moneylog = function(headNode, bodyNode, cb)
{
    moneyLogPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.order = function(headNode, bodyNode, cb)
{
    orderPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.ticket = function(headNode, bodyNode, cb)
{
    ticketPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.relation = function(headNode, bodyNode, cb)
{
    relationPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.print = function(headNode, bodyNode, cb)
{
    printPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.notify = function(headNode, bodyNode, cb)
{
    notifyPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.mongo = function(headNode, bodyNode, cb)
{
    mongoPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.mysql = function(headNode, bodyNode, cb)
{
    mysqlPageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.report = function(headNode, bodyNode, cb)
{
    reportPageControl.handle(headNode, bodyNode, cb);
};


PageControl.prototype.recharge = function(headNode, bodyNode, cb)
{
    rechargePageControl.handle(headNode, bodyNode, cb);
};

PageControl.prototype.sale = function(headNode, bodyNode, cb)
{
    salePageControl.handle(headNode, bodyNode, cb);
};

var pageControl = new PageControl();
module.exports = pageControl;
