// mcp_control/PrintControl.js
var async = require('async');
var esut = require('easy_util');
var log = esut.log;
var digestUtil = esut.digestUtil;
var pageUtil = esut.pageUtil;

var dc = require('mcp_db').dc;

var cons = require('mcp_constants');
var userType = cons.userType;
var msgStatus = cons.msgStatus;
var msgType = cons.msgType;
var ticketPrintQueenStatus = cons.ticketPrintQueenStatus;
var ticketPrintStatus = cons.ticketPrintStatus;
var termStatus = cons.termStatus;
var gameGrade = cons.gameGrade;

var config = require('mcp_config');
var ec = config.ec;
var prop = config.prop;

var service = require("mcp_service");
var digestSer = service.digestSer;
var moneySer = service.moneySer;
var kvSer = service.kvSer;
var ticketSer = service.ticketSer;
var notifySer = service.notifySer;
var termSer = service.termSer;

var PrintControl = function(){
    var self = this;
    self.cmd = {'P01':1, 'P02':2, 'P03':3, 'P04':4,
    'P05':5};
    self.cmdArray = [{},
        {id:1, code:'P01', fromType:prop.digestFromType.FIX, des:"取票"},
        {id:2, code:'P02', fromType:prop.digestFromType.FIX, des:"出票成功"},
        {id:3, code:'P03', fromType:prop.digestFromType.FIX, des:"期次消息"},
        {id:4, code:'P04', fromType:prop.digestFromType.FIX, des:""},
        {id:5, code:'P05', fromType:prop.digestFromType.FIX, des:""}
    ];
};

PrintControl.prototype.handle = function(headNode, bodyStr, userCb)
{
    var self = this;
    async.waterfall([
        //是否是支持的cmd
        function(cb)
        {
            var cmd = self.cmd[headNode.cmd];
            if(cmd == undefined)
            {
                cb(ec.E9000);
            }
            else
            {
                cb(null);
            }
        },
        //校验头的用户类型是否合法
        function(cb)
        {
            var userTypeId = userType[headNode.userType];
            if(userTypeId == undefined)
            {
                cb(ec.E9005);
            }
            else
            {
                cb(null, userTypeId);
            }
        },
        //获得密钥
        function(userTypeId, cb)
        {
            var cmd = self.cmdArray[self.cmd[headNode.cmd]];
            digestSer.getKey({fromType:cmd.fromType, userId:headNode.userId, userType:userTypeId},
            function(err, key){
                cb(err, key);
            });
        },
        //先解密
        function(key, cb)
        {
            log.info(key);
            var decodedBodyStr = digestUtil.check(headNode, key, bodyStr);
            if(decodedBodyStr == null)
            {
                cb(ec.E9003);
                return;
            }
            var bodyNode;
            try {
                bodyNode = JSON.parse(decodedBodyStr);
                headNode.key = key;
            }
            catch (err)
            {
                cb(ec.E9003);
                return;
            }
            cb(null, bodyNode);
        },
        //check the param
        function(bodyNode, cb){
            var method = 'check' + headNode.cmd;
            self[method](null, headNode, bodyNode, function(err){
                cb(err, bodyNode);
            });
        },
        //业务处理
        function(bodyNode, cb){
            var cmd = 'handle' + headNode.cmd;
            self[cmd](null, headNode, bodyNode, cb);
        }
    ], function (err, bodyNode) {
        userCb(err, bodyNode);
    });
};


PrintControl.prototype.checkP01 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

PrintControl.prototype.checkP02 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

PrintControl.prototype.checkP03 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var term = bodyNode.term;
    var passed = false;
    if(term && term.gameCode && term.code)
    {
        if(term.status == termStatus.DRAW)
        {
            //开奖状态需要校验开奖号码
            if(term.wNum)
            {
                passed = true;
            }
            else
            {
                passed = false;
            }
        }
        else
        {
            passed = true;
        }
    }
    if(passed)
    {
        cb(null);
    }
    else
    {
        cb(ec.E2070);
    }
};

PrintControl.prototype.checkP04 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

PrintControl.prototype.checkP05 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

PrintControl.prototype.checkP06 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

PrintControl.prototype.checkP07 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};

PrintControl.prototype.checkP08 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};


PrintControl.prototype.checkP09 = function(user, headNode, bodyNode, cb)
{
    cb(null);
};


/**
 * 取票
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PrintControl.prototype.handleP01 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    var table = dc.mg.get("printqueen");
    pageUtil.parsePost(bodyNode, backBodyNode);
    var now = new Date().getTime();
    var limit = backBodyNode.limit;
    var count = 0;
    var hasNext = true;
    var rst = [];

    async.whilst(
        function() { return count < limit && hasNext},
        function(whileCb) {
            var cond = {printQueenStatus:ticketPrintQueenStatus.WAITING_GET, printId:headNode.userId};
            var sort = {};
            var doc = {$set:{printQueenStatus:ticketPrintQueenStatus.TAKE_AWAY,
                takeTime:now}};
            table.findAndModify(cond, sort, doc, [], function(err, data){
                if(data)
                {
                    data.id = data._id;
                    delete data._id;
                    delete data.version;
                    delete data.status;
                    delete data.printId;
                    rst[rst.length] = data;
                }
                else
                {
                    hasNext = false;
                }
                count++;
                whileCb();
            });
        },
        function(err) {
            backBodyNode.rst = rst;
            cb(err, backBodyNode);
        }
    );
};

/**
 * 出票返回
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PrintControl.prototype.handleP02 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    var objs = bodyNode.rst;
    if(!objs || objs.length == 0)
    {
        cb(ec.E2070, null);
        return;
    }
    var table = dc.mg.get("printqueen");
    var now = new Date().getTime();
    async.each(objs, function(ticket, callback) {
        if(typeof(ticket.id) == "string" ){
            ticket.id = parseInt(ticket.id, 10);
        }
        var cond = {_id:ticket.id};
        var sort = {};
        //如果未打印时间，则使用创建时间
        if(ticket.printTime != undefined && ticket.printTime != null)
        {
            ticket.printTime = dateUtil.toDate(ticket.printTime).getTime();
        }
        if(ticket.cashTime != undefined && ticket.cashTime != null)
        {
            ticket.cashTime = dateUtil.toDate(ticket.cashTime).getTime();
        }
        if(ticket.status == ticketPrintStatus.PRINT_SUCCESS)
        {
            ticketSer.printSuccess(ticket, function(err, backTicket){
                if(err)
                {
                    log.info(err);
                    callback();
                    return;
                }
                backTicket.id = backTicket._id;
                delete backTicket._id;
                delete backTicket.takeTime;
                delete backTicket.version;
                delete backTicket.printQueenStatus;

                backTicket.province = ticket.province;
                backTicket.seq = ticket.seq;
                backTicket.terminal = ticket.terminal;
                backTicket.printTime = ticket.printTime;
                backTicket.cashTime = ticket.cashTime;
                backTicket.prizeTime = ticket.prizeTime;
                backTicket.cashTerminal = ticket.cashTerminal;

                notifySer.saveTicket(backTicket, function(err, data){
                    callback(err);
                });
            });
        }
        else
        {
            var doc = {$set:{status:ticketPrintQueenStatus.PRINT_FAILURE}};
            table.findAndModify(cond, sort, doc, [], function(err, data){
                log.info(data);
                callback();
            });
        }
    }, function(err){
        cb(err, backBodyNode);
    });
};

/**
 * 出票中心通知期次消息，目前只支持通知开奖号码
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PrintControl.prototype.handleP03 = function(user, headNode, bodyNode, cb)
{
    var self = this;
    var backBodyNode = {};
    var term = bodyNode.term;
    var termId = term.gameCode + "_" + term.code;
    if(term.status == termStatus.DRAW)
    {
        var fromTerm = {id:termId, wNum:term.wNum};
        async.waterfall([
            //查看是否有奖级信息
            function(cb)
            {
                var table = dc.main.get("gamegrade");
                var cond = {gameCode:term.gameCode, termCode:term.code};
                var cursor = table.find(cond, {}, []);
                cursor.count(function(err, count){
                    if(count == 0)  //没有奖级信息，从模板导入
                    {
                        var grades = gameGrade.getInfoById(term.gameCode).grades;
                        async.each(grades, function(grade, callback) {
                            var gg = {name:grade.des, gameCode:term.gameCode, termCode:term.code,
                                level:grade.id, bonus:grade.bonus, count:0};
                            gg.id = gg.gameCode + "_" + gg.termCode + "_" + gg.level;
                            table.save(gg, {}, function(err, data){
                                if(err)
                                {
                                    log.info(err);
                                }
                                callback();
                            });
                        }, function(err){
                            cb(err);
                        });
                    }
                    else
                    {
                        cb(null);
                    }
                });
            },
            //更新期次状态
            function (cb) {
                termSer.draw(fromTerm, function (err, data) {
                    cb(err, data);
                });
            },
            //发送期次已经开奖的消息
            function (term, cb) {
                if(term.status == termStatus.DRAW)
                {
                    notifySer.saveTerm(term, cb);
                }
                else
                {
                    cb(null);
                }
            }
        ], function (err, rst) {
            cb(err, backBodyNode);
        });
    }
    else if(term.status == termStatus.NOT_ON_SALE)  //未开售，则更新期次信息
    {
        if(term.gameCode == 'F04')  //如果是内蒙快3，更新期次开售和停售时间
        {
            var cond = {id:term.gameCode + "_" + term.code};
            var data = {
                $set:{
                    openTime:term.openTime,
                    closeTime:term.closeTime
                }
            }
            if(term.bonusTime)  //如果给了开奖时间，则设置开奖时间
            {
                data.$set.bonusTime = term.bonusTime;
            }
            var table = dc.main.get("term");
            table.update(cond, data, {}, function (err, data) {
                cb(err, backBodyNode);
            });
        }
        else
        {
            cb(ec.E2070);
        }
    }
    else
    {
        cb(ec.E2070);
    }
};

/**
 * 删除消息
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PrintControl.prototype.handleP04 = function(user, headNode, bodyNode, cb)
{
};

/**
 * 修改期次
 * @param user
 * @param headNode
 * @param bodyNode
 * @param cb
 */
PrintControl.prototype.handleP05 = function(user, headNode, bodyNode, cb)
{
};

var printControl = new PrintControl();
module.exports = printControl;
